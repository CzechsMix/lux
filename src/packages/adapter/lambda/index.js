import type { AdapterFactory } from '../index';

const lambda: AdapterFactory = () => () => (
  Promise.resolve([
    {},
    {}
  ])
);

export default lambda;
