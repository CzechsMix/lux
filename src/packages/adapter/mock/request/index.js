// @flow
import { parse as parseURL } from 'url';

import qs from 'qs';

import type Logger from '../../../logger';
import type { Request } from '../../../request';
import * as query from '../../utils/query';
import * as method from '../../utils/method';
import { RequestHeaders } from '../../utils/headers';
import type { ObjectMap } from '../../../../interfaces';

type Options = {
  url: string;
  body?: Object;
  method: string;
  logger: Logger;
  headers: ObjectMap<string>;
};

export function create(options: Options): Request {
  const url = parseURL(options.url);
  const params = query.fromObject(qs.parse(url.query));
  const headers = new RequestHeaders(options.headers);

  if (options.body) {
    Object.assign(params, options.body);
  }

  return {
    params,
    headers,
    url: { ...url, params: [] },
    logger: options.logger,
    method: method.resolve(options.method, headers),
    encrypted: url.protocol === 'https:',
    defaultParams: {},
  };
}
