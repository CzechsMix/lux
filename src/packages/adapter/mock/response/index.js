// @flow
import { ResponseHeaders } from '../../utils/headers';
import type Logger from '../../../logger';
import type { Response } from '../../../response';

type Options = {
  logger: Logger;
  resolve?: (data: any) => void;
};

export function create(options: Options): Response {
  const headers = new ResponseHeaders(() => undefined);
  const response = {
    headers,
    end: (body: string) => response.send(body),
    send: (body: string) => {
      if (options.resolve) {
        options.resolve({
          body,
          headers,
          status: response.statusCode,
          statusText: response.statusMessage,
        });
      }
    },
    stats: [],
    logger: options.logger,
    status: (code: number) => {
      response.statusCode = code;
      return response;
    },
    getHeader: key => headers.get(key),
    setHeader: (key, value) => {
      headers.set(key, value);
    },
    removeHeader: key => {
      headers.delete(key);
    },
    statusCode: 200,
    statusMessage: 'OK',
  };

  return response;
}
