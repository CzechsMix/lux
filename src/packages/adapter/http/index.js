// @flow
import { createServer } from 'http';
import type { IncomingMessage, ServerResponse } from 'http';

import { PORT, LUX_CONSOLE } from '../../../constants';
import type { AdapterFactory } from '../index';

import { create as createRequest } from './request';
import { create as createResponse } from './response';

const http: AdapterFactory = app => {
  if (!LUX_CONSOLE) {
    const server = createServer(app.exec);

    server.listen(PORT);
  }

  return (req: IncomingMessage, res: ServerResponse) => (
    createRequest(req, app.logger)
      .then(request => [
        request,
        createResponse(res, app.logger),
      ])
      .catch(err => {
        app.logger.error(err);

        // eslint-disable-next-line no-param-reassign
        res.statusCode = 500;
        res.end(err.stack || err.message);
      })
  );
};

export default http;
