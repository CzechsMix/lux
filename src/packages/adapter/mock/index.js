// @flow
import type { AdapterFactory } from '../index';
import type { ObjectMap } from '../../../interfaces';

import * as request from './request';
import * as response from './response';

type Options = {
  url: string;
  body?: Object;
  method: string;
  headers: ObjectMap<string>;
  resolve?: (data: any) => void;
};

const mock: AdapterFactory = ({ logger }) => (
  ({ url, body, method, headers, resolve }: Options) => (
    Promise.resolve([
      request.create({
        url,
        body,
        method,
        logger,
        headers,
      }),
      response.create({
        logger,
        resolve,
      }),
    ])
  )
);

export default mock;
export { request, response };
