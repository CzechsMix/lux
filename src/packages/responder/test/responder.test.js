// @flow
import Logger from '../../logger';
import { request, response } from '../../adapter/mock';
import noop from '../../../utils/noop';
import * as responder from '../index';

describe('module "responder"', () => {
  describe('#createResponder()', () => {
    let respond;
    let resolve;

    const logger = new Logger({
      level: 'ERROR',
      format: 'text',
      filter: {
        params: [],
      },
      enabled: false,
    });

    beforeEach(() => {
      resolve = jest.fn(noop);
      respond = responder.create(
        request.create({
          logger,
          url: '/health',
          params: {},
          method: 'GET',
          headers: new Map(),
          encrypted: false,
          defaultParams: {},
        }),
        response.create({
          logger,
          resolve,
        })
      );
    });

    afterEach(() => {
      resolve.mockClear();
    });

    it('creates a #respond() function', () => {
      expect(respond).not.toThrow();
      expect(respond).toHaveLength(1);
    });

    describe('#respond()', () => {
      describe('- responding with a string', () => {
        it('works as expected', () => {
          expect(() => respond('Hello World')).not.toThrow();
          expect(resolve.mock.calls).toMatchSnapshot();
        });
      });

      describe('- responding with a number', () => {
        it('works with `204`', () => {
          expect(() => respond(204)).not.toThrow();
          expect(resolve.mock.calls).toMatchSnapshot();
        });

        it('works with `400`', () => {
          expect(() => respond(400)).not.toThrow();
          expect(resolve.mock.calls).toMatchSnapshot();
        });

        it('invalid status codes result in `404`', () => {
          expect(() => respond(1000)).not.toThrow();
          expect(resolve.mock.calls).toMatchSnapshot();
        });
      });

      describe('- responding with a boolean', () => {
        it('works with `true`', () => {
          expect(() => respond(true)).not.toThrow();
          expect(resolve.mock.calls).toMatchSnapshot();
        });

        it('works with `false`', () => {
          expect(() => respond(false)).not.toThrow();
          expect(resolve.mock.calls).toMatchSnapshot();
        });
      });

      describe('- responding with an object', () => {
        it('works with `null`', () => {
          expect(() => respond(null)).not.toThrow();
          expect(resolve.mock.calls).toMatchSnapshot();
        });

        it('works with an object', () => {
          expect(() => respond({ test: true })).not.toThrow();
          expect(resolve.mock.calls).toMatchSnapshot();
        });

        it('works with an array', () => {
          expect(() => respond(['test', true])).not.toThrow();
          expect(resolve.mock.calls).toMatchSnapshot();
        });
      });

      describe('- responding with an error', () => {
        it('works with vanilla errors', () => {
          expect(() => respond(new Error('test'))).not.toThrow();
          expect(resolve.mock.calls).toMatchSnapshot();
        });

        it('works with errors containing a `statusCode` property', () => {
          class ForbiddenError extends Error {
            statusCode = 403;

            constructor() {
              super('Forbidden');
            }
          }

          expect(() => respond(new ForbiddenError())).not.toThrow();
          expect(resolve.mock.calls).toMatchSnapshot();
        });
      });

      describe('- responding with undefined', () => {
        it('works as expected', () => {
          expect(respond).not.toThrow();
          expect(resolve.mock.calls).toMatchSnapshot();
        });
      });
    });
  });
});
