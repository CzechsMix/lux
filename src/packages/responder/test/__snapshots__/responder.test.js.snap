exports[`module "responder" #createResponder() #respond() - responding with a boolean works with \`false\` 1`] = `
Array [
  Array [
    Object {
      "body": "{\"errors\":[{\"status\":\"401\",\"title\":\"Unauthorized\"}],\"jsonapi\":{\"version\":\"1.0\"}}",
      "headers": Array [
        Array [
          "content-type",
          "application/vnd.api+json",
        ],
      ],
      "status": 401,
      "statusText": "OK",
    },
  ],
]
`;

exports[`module "responder" #createResponder() #respond() - responding with a boolean works with \`true\` 1`] = `
Array [
  Array [
    Object {
      "body": "undefined",
      "headers": Array [],
      "status": 204,
      "statusText": "OK",
    },
  ],
]
`;

exports[`module "responder" #createResponder() #respond() - responding with a number invalid status codes result in \`404\` 1`] = `
Array [
  Array [
    Object {
      "body": "{\"errors\":[{\"status\":\"404\",\"title\":\"Not Found\"}],\"jsonapi\":{\"version\":\"1.0\"}}",
      "headers": Array [
        Array [
          "content-type",
          "application/vnd.api+json",
        ],
      ],
      "status": 404,
      "statusText": "OK",
    },
  ],
]
`;

exports[`module "responder" #createResponder() #respond() - responding with a number works with \`204\` 1`] = `
Array [
  Array [
    Object {
      "body": "",
      "headers": Array [],
      "status": 204,
      "statusText": "OK",
    },
  ],
]
`;

exports[`module "responder" #createResponder() #respond() - responding with a number works with \`400\` 1`] = `
Array [
  Array [
    Object {
      "body": "{\"errors\":[{\"status\":\"400\",\"title\":\"Bad Request\"}],\"jsonapi\":{\"version\":\"1.0\"}}",
      "headers": Array [
        Array [
          "content-type",
          "application/vnd.api+json",
        ],
      ],
      "status": 400,
      "statusText": "OK",
    },
  ],
]
`;

exports[`module "responder" #createResponder() #respond() - responding with a string works as expected 1`] = `
Array [
  Array [
    Object {
      "body": "Hello World",
      "headers": Array [
        Array [
          "content-type",
          "text/plain",
        ],
      ],
      "status": 200,
      "statusText": "OK",
    },
  ],
]
`;

exports[`module "responder" #createResponder() #respond() - responding with an error works with errors containing a \`statusCode\` property 1`] = `
Array [
  Array [
    Object {
      "body": "{\"errors\":[{\"status\":\"403\",\"title\":\"Forbidden\",\"detail\":\"Forbidden\"}],\"jsonapi\":{\"version\":\"1.0\"}}",
      "headers": Array [
        Array [
          "content-type",
          "application/vnd.api+json",
        ],
      ],
      "status": 403,
      "statusText": "OK",
    },
  ],
]
`;

exports[`module "responder" #createResponder() #respond() - responding with an error works with vanilla errors 1`] = `
Array [
  Array [
    Object {
      "body": "{\"errors\":[{\"status\":\"500\",\"title\":\"Internal Server Error\",\"detail\":\"test\"}],\"jsonapi\":{\"version\":\"1.0\"}}",
      "headers": Array [
        Array [
          "content-type",
          "application/vnd.api+json",
        ],
      ],
      "status": 500,
      "statusText": "OK",
    },
  ],
]
`;

exports[`module "responder" #createResponder() #respond() - responding with an object works with \`null\` 1`] = `
Array [
  Array [
    Object {
      "body": "{\"errors\":[{\"status\":\"404\",\"title\":\"Not Found\"}],\"jsonapi\":{\"version\":\"1.0\"}}",
      "headers": Array [
        Array [
          "content-type",
          "application/vnd.api+json",
        ],
      ],
      "status": 404,
      "statusText": "OK",
    },
  ],
]
`;

exports[`module "responder" #createResponder() #respond() - responding with an object works with an array 1`] = `
Array [
  Array [
    Object {
      "body": "[\"test\",true]",
      "headers": Array [
        Array [
          "content-type",
          "application/vnd.api+json",
        ],
      ],
      "status": 200,
      "statusText": "OK",
    },
  ],
]
`;

exports[`module "responder" #createResponder() #respond() - responding with an object works with an object 1`] = `
Array [
  Array [
    Object {
      "body": "{\"test\":true}",
      "headers": Array [
        Array [
          "content-type",
          "application/vnd.api+json",
        ],
      ],
      "status": 200,
      "statusText": "OK",
    },
  ],
]
`;

exports[`module "responder" #createResponder() #respond() - responding with undefined works as expected 1`] = `
Array [
  Array [
    Object {
      "body": "{\"errors\":[{\"status\":\"404\",\"title\":\"Not Found\"}],\"jsonapi\":{\"version\":\"1.0\"}}",
      "headers": Array [
        Array [
          "content-type",
          "application/vnd.api+json",
        ],
      ],
      "status": 404,
      "statusText": "OK",
    },
  ],
]
`;
